### PR TITLE
[Good First Issue] Verify baichuan2-7b-chat with GenAI text_generation

### PR DESCRIPTION
**Pull request by mengbingrock**
Verified baichuan2-7b-chat with GenAI text_generation, added it to Github workflow and README.

This PR should be merged after a change to optimum-intel is made: [273](https://github.com/openvinotoolkit/openvino.genai/issues/273#issuecomment-2017885879)

### CODE REVIEW

**Comment by mengbingrock** [comment]
Not sure why the CI failed, it could run locally:
```
convert_tokenizer ./Baichuan2-7B-Chat/pytorch/dldt/FP16/ --output ./Baichuan2-7B-Chat/pytorch/dldt/FP16/ --with-detokenizer --trust-remote-code
Loading Huggingface Tokenizer...
Converting Huggingface Tokenizer to OpenVINO...
Saved OpenVINO Tokenizer: Baichuan2-7B-Chat/pytorch/dldt/FP16/openvino_tokenizer.xml, Baichuan2-7B-Chat/pytorch/dldt/FP16/openvino_tokenizer.bin
Saved OpenVINO Detokenizer: Baichuan2-7B-Chat/pytorch/dldt/FP16/openvino_detokenizer.xml, Baichuan2-7B-Chat/pytorch/dldt/FP16/openvino_detokenizer.bin
```

**Comment by mengbingrock** [comment]
> Not sure why the CI failed, it could run locally:
> 
> ```
> convert_tokenizer ./Baichuan2-7B-Chat/pytorch/dldt/FP16/ --output ./Baichuan2-7B-Chat/pytorch/dldt/FP16/ --with-detokenizer --trust-remote-code
> Loading Huggingface Tokenizer...
> Converting Huggingface Tokenizer to OpenVINO...
> Saved OpenVINO Tokenizer: Baichuan2-7B-Chat/pytorch/dldt/FP16/openvino_tokenizer.xml, Baichuan2-7B-Chat/pytorch/dldt/FP16/openvino_tokenizer.bin
> Saved OpenVINO Detokenizer: Baichuan2-7B-Chat/pytorch/dldt/FP16/openvino_detokenizer.xml, Baichuan2-7B-Chat/pytorch/dldt/FP16/openvino_detokenizer.bin
> ```

Just noticed that the build step failed, so that folder is not existed:

```
OpenVINO Tokenizer version is not compatible with OpenVINO version. Installed OpenVINO version: 2024.1.0,OpenVINO Tokenizers requires 2024.0.0. OpenVINO Tokenizers models will not be added during export.
[ INFO ] openvino runtime version: 2024.1.0-14645-e6dc0865128
A new version of the following files was downloaded from https://huggingface.co/baichuan-inc/Baichuan2-7B-Chat:
- configuration_baichuan.py
. Make sure to double-check they do not contain any added malicious code. To avoid downloading new versions of the code file, you can pin a revision.
Traceback (most recent call last):
  File "./llm_bench/python/convert.py", line 1411, in <module>
    main()
  File "./llm_bench/python/convert.py", line 1408, in main
    converter(args)
  File "./llm_bench/python/convert.py", line 1215, in convert_baichaun
    model = AutoModelForCausalLM.from_pretrained(args.model_id, trust_remote_code=True, **model_kwargs)
  File "/opt/hostedtoolcache/Python/3.8.18/x64/lib/python3.8/site-packages/transformers/models/auto/auto_factory.py", line 548, in from_pretrained
    model_class = get_class_from_dynamic_module(
  File "/opt/hostedtoolcache/Python/3.8.18/x64/lib/python3.8/site-packages/transformers/dynamic_module_utils.py", line 488, in get_class_from_dynamic_module
    final_module = get_cached_module_file(
  File "/opt/hostedtoolcache/Python/3.8.18/x64/lib/python3.8/site-packages/transformers/dynamic_module_utils.py", line 352, in get_cached_module_file
    get_cached_module_file(
  File "/opt/hostedtoolcache/Python/3.8.18/x64/lib/python3.8/site-packages/transformers/dynamic_module_utils.py", line 314, in get_cached_module_file
    modules_needed = check_imports(resolved_module_file)
  File "/opt/hostedtoolcache/Python/3.8.18/x64/lib/python3.8/site-packages/transformers/dynamic_module_utils.py", line 180, in check_imports
    raise ImportError(
ImportError: This modeling file requires the following packages that were not found in your environment: bitsandbytes. Run `pip install bitsandbytes`
```

Need to find a way to solve above issue

**Comment by mengbingrock** [comment]
Hello @pavel-esir , I've added pip install bitsandbytes step before  running this model. Could we try CI again to see if this time it works?  Thanks a lot!

**Comment by mengbingrock** [comment]
I've checked the new error in CI, this is expected. We need to wait for the change to intel-optimum to be merged. [https://github.com/openvinotoolkit/openvino.genai/issues/273#issuecomment-2017885879](https://github.com/openvinotoolkit/openvino.genai/issues/273#issuecomment-2017885879)

File: .github/workflows/causal_lm_cpp.yml [hunk 1]
```diff
@@ -218,6 +218,32 @@ jobs:
...
+          source ./ov/setupvars.sh
+          python -m pip install --upgrade-strategy eager "optimum>=1.14" "bitsandbytes>=0.43.0" -r ./llm_bench/python/requirements.txt ./thirdparty/openvino_tokenizers/[transformers] --extra-index-url https://download.pytorch.org/whl/cpu && python ./llm_bench/python/convert.py --model_id baichuan-inc/Baichuan2-7B-Chat --output_dir ./Baichuan2-7B-Chat/ --precision FP16 &
+          cmake -DCMAKE_BUILD_TYPE=Release -S ./text_generation/causal_lm/cpp/ -B ./build/
+          cmake --build ./build/ --config Release -j
+          wait
+      - name: Run
+        run: |
+          source ./ov/setupvars.sh
+          convert_tokenizer ./Baichuan2-7B-Chat/pytorch/dldt/FP16/ --output ./Baichuan2-7B-Chat/pytorch/dldt/FP16/ --with-detokenizer --trust-remote-code
+          timeout 50s ./build/beam_search_causal_lm ./Baichuan2-7B-Chat/pytorch/dldt/FP16/ "你好！" > ./pred_baichuan2.txt
```
Comment (ilya-lavrenov) [inline .github/workflows/causal_lm_cpp.yml:246]: could you please also add comparison with Hugging Face?
Example is https://github.com/openvinotoolkit/openvino.genai/blob/f973f6287ed5d9decf1bd3e58e88e6c813e24378/.github/workflows/causal_lm_cpp.yml#L67-L80

Comment (mengbingrock) [reply .github/workflows/causal_lm_cpp.yml]: Hi @ilya-lavrenov , Thanks for your review. I've added the compare logic. In my code it replaced `transformers.LlamaTokenizer` with `transformers.AutoTokenizer` and `transformers.LlamaForCausalLM.from_pretrained` with `transformers.AutoModelForCausalLM.from_pretrained`. I hope that looks good to you.

File: .github/workflows/causal_lm_cpp.yml [hunk 2]
```diff
@@ -244,6 +244,47 @@ jobs:
...
+          python-version: 3.8
+      - name: Install OpenVINO
+        run: |
+          mkdir ./ov/
+          curl https://storage.openvinotoolkit.org/repositories/openvino/packages/nightly/2024.1.0-14645-e6dc0865128/l_openvino_toolkit_ubuntu20_2024.1.0.dev20240304_x86_64.tgz | tar --directory ./ov/ --strip-components 1 -xz
+          sudo ./ov/install_dependencies/install_openvino_dependencies.sh
+      - name: Download, convert and build
+        run: |
+          source ./ov/setupvars.sh
+          python -m pip install --upgrade-strategy eager "optimum>=1.14" -r ./llm_bench/python/requirements.txt ./thirdparty/openvino_tokenizers/[transformers] --extra-index-url https://download.pytorch.org/whl/cpu && python ./llm_bench/python/convert.py --model_id baichuan-inc/Baichuan2-7B-Chat --output_dir ./Baichuan2-7B-Chat/ --precision FP16 &
```
Comment (ilya-lavrenov) [inline .github/workflows/causal_lm_cpp.yml:264]: let's install latest `main` https://github.com/huggingface/optimum-intel.git which contains fix for `baichuan2-7b-chat` model

File: .github/workflows/causal_lm_cpp.yml [hunk 3]
```diff
@@ -265,6 +265,48 @@ jobs:
...
+          python -c "
+          import transformers
+          with open('pred_baichuan2.txt', 'r') as file:
+            predictions = file.read()
+
+          tokenizer = transformers.AutoTokenizer.from_pretrained('baichuan-inc/Baichuan2-7B-Chat',trust_remote_code=True)
+          tokenized = tokenizer('69', return_tensors='pt')
+          for beam in transformers.AutoModelForCausalLM.from_pretrained('baichuan-inc/Baichuan2-7B-Chat',trust_remote_code=True).generate(**tokenized, num_beam_groups=3, num_beams=15, num_return_sequences=15, diversity_penalty=1.0, max_new_tokens=20, early_stopping=False, length_penalty=1.0, no_repeat_ngram_size=9**9, do_sample=False):
+              ref = tokenizer.decode(beam[tokenized['input_ids'].numel():], skip_special_tokens=True)
+              idx = predictions.find(ref)
```
Comment (ilya-lavrenov) [inline .github/workflows/causal_lm_cpp.yml:303]: now, we have error on this line:
```
RuntimeError: Missing ref='69696969696969696969' from predictions
```

Comment (mengbingrock) [reply .github/workflows/causal_lm_cpp.yml:303]: yes, same happened locally. I'm not sure if it is the problem. Could it due to the random nature of generating? I'm studying into it.

Comment (ilya-lavrenov) [reply .github/workflows/causal_lm_cpp.yml:303]: image
maybe we need to address these warnings? let's try to fully override [GenerationConfig](https://huggingface.co/docs/transformers/main_classes/text_generation#transformers.GenerationConfig) for the model

Comment (mengbingrock) [reply .github/workflows/causal_lm_cpp.yml:303]: Great suggestion, these warning could be the clue. I'll try to understand these parameters and make change.

**Comment by andrei-kochin** [comment]
@mengbingrock are you still working on this? if so could you please resolve conflicts with the latest master?

**Comment by mengbingrock** [comment]
> @mengbingrock are you still working on this? if so could you please resolve conflicts with the latest master?

I could continue working in this, first I'll resolve the conflicts with the latest master.

**Comment by Wovchena** [comment]
@mengbingrock any updates here?

**Pull request closed without merging**